hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestReplicationManagerScenarios.java

```diff
@@ -0,0 +1,783 @@
+    for (SCMCommandProto.Type type : SCMCommandProto.Type.values()) {
+      countMap.put(type, 0);
+    }
+    Mockito.when(
+        nodeManager.getTotalDatanodeCommandCounts(any(DatanodeDetails.class),
+            any(SCMCommandProto.Type.class), any(SCMCommandProto.Type.class)))
+        .thenReturn(countMap);
+
+    // Ensure that RM will run when asked.
+    Mockito.when(scmContext.isLeaderReady()).thenReturn(true);
+    Mockito.when(scmContext.isInSafeMode()).thenReturn(false);
+    containerReplicaMap = new HashMap<>();
+    containerInfoSet = new HashSet<>();
+    repReport = new ReplicationManagerReport();
+    repQueue = new ReplicationQueue();
+    ORIGINS.clear();
+    DATANODE_ALIASES.clear();
+    NODE_STATUS_MAP.clear();
+  }
+
+  private ReplicationManager createReplicationManager() throws IOException {
+    return new ReplicationManager(
+        configuration,
+        containerManager,
+        ratisPlacementPolicy,
+        ecPlacementPolicy,
+        eventPublisher,
+        scmContext,
+        nodeManager,
+        clock,
+        legacyReplicationManager,
+        containerReplicaPendingOps) {
+      @Override
+      protected void startSubServices() {
+        // do not start any threads for processing
+      }
+    };
+  }
+
+  protected static UUID getOrCreateOrigin(String origin) {
+    return ORIGINS.computeIfAbsent(origin, (k) -> UUID.randomUUID());
+  }
+
+  private static Stream<Scenario> getTestScenarios() {
+    return TEST_SCENARIOS.stream();
+  }
+
+  private void loadPendingOps(ContainerInfo container, Scenario scenario) {
+    for (PendingReplica r : scenario.getPendingReplicas()) {
+      if (r.getType() == ContainerReplicaOp.PendingOpType.ADD) {
+        containerReplicaPendingOps.scheduleAddReplica(container.containerID(), r.getDatanodeDetails(),
+            r.getReplicaIndex(), Long.MAX_VALUE);
+      } else if (r.getType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        containerReplicaPendingOps.scheduleDeleteReplica(container.containerID(), r.getDatanodeDetails(),
+            r.getReplicaIndex(), Long.MAX_VALUE);
+      }
+    }
+  }
+
+  @ParameterizedTest
+  @MethodSource("getTestScenarios")
+  public void testAllScenarios(Scenario scenario) throws IOException {
+    ReplicationManager.ReplicationManagerConfiguration conf =
+        new ReplicationManager.ReplicationManagerConfiguration();
+    conf.setMaintenanceRemainingRedundancy(scenario.getEcMaintenanceRedundancy());
+    conf.setMaintenanceReplicaMinimum(scenario.getRatisMaintenanceMinimum());
+    configuration.setFromObject(conf);
+    replicationManager = createReplicationManager();
+
+    ContainerInfo containerInfo = scenario.buildContainerInfo();
+    loadPendingOps(containerInfo, scenario);
+
+    Set<ContainerReplica> replicas = new HashSet<>();
+    for (TestReplica replica : scenario.getReplicas()) {
+      replicas.add(replica.buildContainerReplica());
```

Comment:
The back and forth between the inner and outer classes with conversion methods like `buildContainerReplica` looks a bit strange. Can we make those methods part of the outer class so the inner classes just need to be concerned with holding state, and the outer class does translation into the form it wants to work with?

Reply:
The inner class here represents a replica, so having `buildContainerReplica` on it makes sense to me. Having it on the scenario might be somewhat strange too. I guess Scenario could have getContainerReplicas, where it iterates the replicas internally, but all that does is move this for loop from here to there.

Being that these methods are only used inside this one test method and the classes are localized to this class for testing, I am not sure its worth over thinking the access methods. It seems reasonably clear to me as it stands.